ndk/src/midi.rs

```diff
@@ -0,0 +1,300 @@
+        let num_bytes_sent = unsafe {
+            ffi::AMidiInputPort_sendWithTimestamp(
+                self.ptr.as_ptr(),
+                buffer.as_ptr(),
+                buffer.len() as ffi::size_t,
+                timestamp,
+            )
+        };
+        if num_bytes_sent >= 0 {
+            Ok(num_bytes_sent as usize)
+        } else {
+            Err(
+                NdkMediaError::from_status(ffi::media_status_t(num_bytes_sent as c_int))
+                    .unwrap_err(),
+            )
+        }
+    }
+}
+
+impl<'a> Drop for MidiInputPort<'a> {
+    fn drop(&mut self) {
+        unsafe { ffi::AMidiInputPort_close(self.ptr.as_ptr()) };
+    }
+}
+
+pub struct MidiOutputPort<'a> {
+    ptr: NonNull<ffi::AMidiOutputPort>,
+    _marker: PhantomData<&'a MidiDevice>,
+}
+
+impl<'a> fmt::Debug for MidiOutputPort<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MidiOutputPort")
+            .field("inner", &self.ptr)
+            .finish()
+    }
+}
+
+impl<'a> MidiOutputPort<'a> {
+    /// Assumes ownership of `ptr`
+    ///
+    /// # Safety
+    /// `ptr` must be a valid pointer to an Android [`ffi::AMidiOutputPort`].
+    pub unsafe fn from_ptr(ptr: NonNull<ffi::AMidiOutputPort>) -> Self {
+        Self {
+            ptr,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn ptr(&self) -> NonNull<ffi::AMidiOutputPort> {
+        self.ptr
+    }
+
+    /// Receives the next pending MIDI message.
+    ///
+    /// To retrieve all pending messages, the client should repeatedly call this method until it
+    /// returns [`Ok(MidiOpcode::NoMessage)`].
+    ///
+    /// Note that this is a non-blocking call. If there are no Midi messages are available, the
+    /// function returns [`Ok(MidiOpcode::NoMessage)`] immediately (for 0 messages received).
+    pub fn receive(&self, buffer: &mut [u8]) -> Result<MidiOpcode> {
+        let mut opcode = 0i32;
+        let mut timestamp = 0i64;
+        let mut num_bytes_received: ffi::size_t = 0;
+        let num_messages_received = unsafe {
+            ffi::AMidiOutputPort_receive(
+                self.ptr.as_ptr(),
+                &mut opcode,
+                buffer.as_mut_ptr(),
+                buffer.len() as ffi::size_t,
+                &mut num_bytes_received,
+                &mut timestamp,
+            )
+        };
+
+        match num_messages_received {
+            r if r < 0 => {
+                Err(NdkMediaError::from_status(ffi::media_status_t(r as c_int)).unwrap_err())
+            }
+            0 => Ok(MidiOpcode::NoMessage),
+            1 => match opcode as c_uint {
+                ffi::AMIDI_OPCODE_DATA => Ok(MidiOpcode::Data {
+                    length: num_bytes_received as usize,
+                    timestamp,
+                }),
+                ffi::AMIDI_OPCODE_FLUSH => Ok(MidiOpcode::Flush),
+                _ => unreachable!("Opcode is {}", opcode),
```

Comment:
Nit: if you can call that `Unrecognized opcode {}` that makes it a bit more clear :)

Reply:
![image](https://user-images.githubusercontent.com/17005454/194055158-d333b14a-4051-4300-b8ba-da3fa36eb6ba.png)
Fixed!